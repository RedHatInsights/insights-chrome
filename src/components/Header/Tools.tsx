/* eslint-disable @typescript-eslint/ban-ts-comment */
import React, { memo, useContext, useEffect, useState } from 'react';
import { useAtom, useAtomValue, useSetAtom } from 'jotai';
import { Button } from '@patternfly/react-core/dist/dynamic/components/Button';
import { Divider } from '@patternfly/react-core/dist/dynamic/components/Divider';
import { DropdownItem } from '@patternfly/react-core/dist/dynamic/components/Dropdown';
import { NotificationBadge } from '@patternfly/react-core/dist/dynamic/components/NotificationBadge';
import { Switch } from '@patternfly/react-core/dist/dynamic/components/Switch';
import { ToolbarItem } from '@patternfly/react-core/dist/dynamic/components/Toolbar';
import { Tooltip } from '@patternfly/react-core/dist/dynamic/components/Tooltip';
import QuestionCircleIcon from '@patternfly/react-icons/dist/dynamic/icons/question-circle-icon';
import CogIcon from '@patternfly/react-icons/dist/dynamic/icons/cog-icon';
import RedhatIcon from '@patternfly/react-icons/dist/dynamic/icons/redhat-icon';
import UserToggle from './UserToggle';
import ToolbarToggle from './ToolbarToggle';
import SettingsToggle, { SettingsToggleDropdownGroup } from './SettingsToggle';
import cookie from 'js-cookie';
import { ITLess, getSection } from '../../utils/common';
import { useIntl } from 'react-intl';
import { useFlag } from '@unleash/proxy-client-react';
import messages from '../../locales/Messages';
import { createSupportCase } from '../../utils/createCase';
import BellIcon from '@patternfly/react-icons/dist/dynamic/icons/bell-icon';
import ChromeAuthContext from '../../auth/ChromeAuthContext';
import { isPreviewAtom, togglePreviewWithCheckAtom } from '../../state/atoms/releaseAtom';
import { notificationDrawerExpandedAtom, unreadNotificationsAtom } from '../../state/atoms/notificationDrawerAtom';
import useSupportCaseData from '../../hooks/useSupportCaseData';

const isITLessEnv = ITLess();

const InternalButton = () => (
  <Button
    variant="plain"
    aria-label="Go to internal tools"
    widget-type="InternalButton"
    className="chr-c-toolbar__button-internal"
    href={`${document.baseURI}internal`}
    component="a"
  >
    <RedhatIcon />
  </Button>
);

type ExpandedSettingsButtonProps = {
  settingsMenuDropdownGroups: SettingsToggleDropdownGroup[];
};

const ExpandedSettingsButton = ({ settingsMenuDropdownGroups }: ExpandedSettingsButtonProps) => (
  <Tooltip aria="none" aria-live="polite" content={'Settings'} flipBehavior={['bottom']} className="tooltip-inner-settings-cy">
    <SettingsToggle
      key="Settings menu"
      icon={() => <CogIcon />}
      id="SettingsMenu"
      ariaLabel="Settings menu"
      ouiaId="chrome-settings"
      hasToggleIndicator={null}
      widget-type="SettingsMenu"
      dropdownItems={settingsMenuDropdownGroups}
      className="tooltip-button-settings-cy"
    />
  </Tooltip>
);

const Tools = () => {
  const [{ isDemoAcc, isInternal, isRhosakEntitled }, setState] = useState({
    isInternal: true,
    isRhosakEntitled: false,
    isDemoAcc: false,
  });
  const isPreview = useAtomValue(isPreviewAtom);
  const togglePreviewWithCheck = useSetAtom(togglePreviewWithCheckAtom);
  const enableIntegrations = useFlag('platform.sources.integrations');
  const workspacesEnabled = useFlag('platform.rbac.workspaces');
  const enableGlobalLearningResourcesPage = useFlag('platform.learning-resources.global-learning-resources');
  const { user, token } = useContext(ChromeAuthContext);
  const unreadNotifications = useAtomValue(unreadNotificationsAtom);
  const [isNotificationDrawerExpanded, toggleNotifications] = useAtom(notificationDrawerExpandedAtom);
  const intl = useIntl();
  const isOrgAdmin = user?.identity?.user?.is_org_admin;
  const settingsPath = isITLessEnv ? `/settings/my-user-access` : enableIntegrations ? `/settings/integrations` : '/settings/sources';
  const identityAndAccessManagmentPath = isOrgAdmin
    ? `/iam/${workspacesEnabled ? 'access-management' : 'user-access'}/overview`
    : '/iam/my-user-access';
  const betaSwitcherTitle = `${isPreview ? intl.formatMessage(messages.stopUsing) : intl.formatMessage(messages.use)} ${intl.formatMessage(
    messages.betaRelease
  )}`;

  const isNotificationsEnabled = useFlag('platform.chrome.notifications-drawer');

  /* list out the items for the settings menu */
  const settingsMenuDropdownGroups = [
    {
      items: [
        {
          ouiaId: 'PreviewSwitcher',
          title: `${isPreview ? 'Exit' : 'Enable'} "Preview" mode`,
          url: '#',
          onClick: () => togglePreviewWithCheck(),
        },
      ],
    },
    {
      title: 'Settings',
      items: [
        {
          url: '/settings/integrations',
          title: 'Integrations',
        },
        {
          url: '/settings/notifications',
          title: 'Notifications',
        },
      ],
    },
    {
      title: 'Identity and Access Management',
      items: [
        {
          url: identityAndAccessManagmentPath,
          title: isOrgAdmin ? (workspacesEnabled ? 'Acess management' : 'User Access') : 'My User Access',
        },
        {
          url: '/iam/authentication-policy/authentication-factors',
          title: 'Authentication Policy',
        },
        {
          url: '/iam/service-accounts',
          title: 'Service Accounts',
        },
      ],
    },
  ];

  useEffect(() => {
    if (user) {
      setState({
        isInternal: !!user?.identity?.user?.is_internal,
        isRhosakEntitled: !!user?.entitlements?.rhosak?.is_entitled,
        isDemoAcc: user?.identity?.user?.username === 'insights-demo-2021',
      });
    }
  }, [user]);
  const supportCaseData = useSupportCaseData();

  const supportOptionsUrl = () => {
    return isITLessEnv ? 'https://redhatgov.servicenowservices.com/css' : 'https://access.redhat.com/support';
  };

  /* list out the items for the about menu */
  const aboutMenuDropdownItems = [
    {
      title: intl.formatMessage(messages.apiDocumentation),
      onClick: () => window.open('https://developers.redhat.com/api-catalog/', '_blank'),
      isHidden: isITLessEnv,
    },
    {
      title: intl.formatMessage(messages.openSupportCase),
      onClick: () => createSupportCase(user.identity, token, isPreview, { supportCaseData }),
      isDisabled: window.location.href.includes('/application-services') && !isRhosakEntitled,
      isHidden: isITLessEnv,
    },
    {
      title: intl.formatMessage(messages.statusPage),
      onClick: () => window.open('https://status.redhat.com/', '_blank'),
      isHidden: isITLessEnv,
    },
    {
      title: intl.formatMessage(messages.supportOptions),
      onClick: () => (window.location.href = supportOptionsUrl()),
    },
    {
      title: intl.formatMessage(messages.insightsRhelDocumentation),
      onClick: () => window.open('https://docs.redhat.com/en/documentation/red_hat_insights', '_blank'),
      isHidden: getSection() !== 'insights' || isITLessEnv,
    },
    {
      title: intl.formatMessage(messages.demoMode),
      onClick: () => cookie.set('cs_demo', 'true') && window.location.reload(),
      isHidden: !isDemoAcc,
    },
    ...(enableGlobalLearningResourcesPage
      ? [
          {
            title: intl.formatMessage(messages.globalLearningResourcesPage),
            onClick: () => window.open('/staging/global-learning-resources-page', '_blank'),
          },
        ]
      : []),
  ];

  /* Combine aboutMenuItems with a settings link on mobile */
  const mobileDropdownItems = [
    { title: 'separator' },
    {
      url: settingsPath,
      title: 'Settings',
      target: '_self',
    },
    {
      title: betaSwitcherTitle,
      onClick: () => togglePreviewWithCheck(),
    },
    { title: 'separator' },
    ...aboutMenuDropdownItems,
  ];

  /* QuestionMark icon that should be used for "help/support" things */
  const AboutButton = () => (
    <Tooltip aria="none" aria-live="polite" content={'Help'} flipBehavior={['bottom']} className="tooltip-inner-help-cy">
      <ToolbarToggle
        key="Help menu"
        icon={QuestionCircleIcon}
        id="HelpMenu"
        ouiaId="chrome-help"
        ariaLabel="Help menu"
        hasToggleIndicator={null}
        dropdownItems={aboutMenuDropdownItems}
        className="tooltip-button-help-cy"
      />
    </Tooltip>
  );

  const ThemeToggle = () => {
    const [darkmode, setDarkmode] = useState(false);
    return (
      <Switch
        id="no-label-switch-on"
        isChecked={darkmode || false}
        aria-label="Dark mode switch"
        onChange={() => {
          setDarkmode(!darkmode);
          document.body.classList.contains('pf-theme-dark')
            ? document.body.classList.remove('pf-theme-dark')
            : document.body.classList.add('pf-theme-dark');
        }}
      />
    );
  };

  return (
    <>
      {isNotificationsEnabled && (
        <ToolbarItem className="pf-v5-u-mx-0">
          <Tooltip aria="none" aria-live="polite" content={'Notifications'} flipBehavior={['bottom']} className="tooltip-inner-settings-cy">
            <NotificationBadge
              className="chr-c-notification-badge"
              variant={unreadNotifications ? 'unread' : 'read'}
              onClick={() => toggleNotifications((prev) => !prev)}
              aria-label="Notifications"
              isExpanded={isNotificationDrawerExpanded}
            >
              <BellIcon />
            </NotificationBadge>
          </Tooltip>
        </ToolbarItem>
      )}
      {localStorage.getItem('chrome:darkmode') === 'true' && (
        <ToolbarItem>
          <ThemeToggle />
        </ToolbarItem>
      )}
      {isInternal && !ITLess() && (
        <ToolbarItem className="pf-v5-u-mr-0">
          <Tooltip aria="none" aria-live="polite" content={'Internal'} flipBehavior={['bottom']}>
            <InternalButton />
          </Tooltip>
        </ToolbarItem>
      )}
      <ToolbarItem className="pf-v5-u-mr-0" visibility={{ default: 'hidden', md: 'visible' }}>
        <ExpandedSettingsButton settingsMenuDropdownGroups={settingsMenuDropdownGroups} />
      </ToolbarItem>
      <ToolbarItem className="pf-v5-u-mr-0" visibility={{ default: 'hidden', md: 'visible' }}>
        <AboutButton />
      </ToolbarItem>
      <ToolbarItem className="pf-v5-u-mr-0" visibility={{ default: 'hidden', lg: 'visible' }}>
        <UserToggle />
      </ToolbarItem>
      {/* Collapse tools and user dropdown to kebab on small screens  */}

      <ToolbarItem visibility={{ lg: 'hidden' }}>
        <Tooltip aria="none" aria-live="polite" content={'More options'} flipBehavior={['bottom']}>
          <UserToggle
            isSmall
            extraItems={mobileDropdownItems.map((action, key) => (
              <React.Fragment key={key}>
                {action.title === 'separator' ? (
                  <Divider component="li" />
                ) : (
                  <DropdownItem
                    {...(action.onClick
                      ? {
                          component: 'button',
                          onClick: action.onClick,
                        }
                      : {
                          href: action.url,
                          component: 'a',
                          target: '_blank',
                          rel: 'noopener noreferrer',
                        })}
                  >
                    {action.title}
                  </DropdownItem>
                )}
              </React.Fragment>
            ))}
          />
        </Tooltip>
      </ToolbarItem>
    </>
  );
};

export default memo(Tools);
