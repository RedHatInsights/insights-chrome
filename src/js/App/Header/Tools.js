import React, { Component } from 'react';
import { Button } from '@patternfly/react-core';
import { Toolbar, ToolbarGroup, ToolbarItem } from '@patternfly/react-core';
import { DropdownItem } from '@patternfly/react-core';
import { QuestionCircleIcon, CogIcon } from '@patternfly/react-icons';

import UserToggle from './UserToggle';
import UserIcon from './UserIcon';
import ToolbarToggle from './ToolbarToggle';
import InsightsAbout from './InsightsAbout';

const aboutButton = {
    title: 'FAQ',
    icon: QuestionCircleIcon,
    widget: 'InsightsFAQ',
    items: [
        {
            title: 'Help & assistance',
            url: 'https://access.redhat.com/help/'
        }, {
            title: 'Customer support',
            url: 'https://access.redhat.com/support'
        }, {
            title: 'Training',
            url: 'https://www.redhat.com/en/services/training-and-certification'
        }, {
            title: 'API Documentation',
            onClick: () => window.location.href = `${document.baseURI}docs/api`
        }, {
            title: 'About'
        }
    ]
};

const actions = [
    aboutButton
];

const settingsButton = (
    <ToolbarItem>
        <Button variant="plain"
            aria-label="Go to settings"
            widget-type='SettingsButton'
            onClick={() => window.location.href = `${document.baseURI}settings/`}>
            <CogIcon/>
        </Button>
    </ToolbarItem>
);

class Tools extends Component {
    constructor(props) {
        super(props);
        this.onModalToggle = this.onModalToggle.bind(this);
        aboutButton.items[4].onClick = this.onModalToggle.bind(this);
        this.state = {
            isModalOpen: false
        };
    }
    onModalToggle() {
        this.setState({
            isModalOpen: !this.state.isModalOpen
        });
    }

    render() {
        const { isModalOpen } = this.state;
        return (
            <div className="pf-l-page__header-tools pf-c-page__header-tools" widget-type="InsightsToolbar">
                <Toolbar>
                    <ToolbarGroup className='pf-u-mr-0 pf-u-mr-lg-on-lg'>
                        { settingsButton }
                        { actions.map((oneItem, key) => (
                            oneItem.items ?
                                <ToolbarToggle key={key} icon={oneItem.icon} dropdownItems={oneItem.items} /> :
                                <ToolbarItem key={key} data-key={key}>
                                    <Button
                                        key={key}
                                        variant="plain"
                                        aria-label={`Overflow ${oneItem.title}`}
                                        widget-type={oneItem.widget}
                                        onClick={event => oneItem.onClick && oneItem.onClick(event)}
                                    >
                                        <oneItem.icon />
                                    </Button>
                                </ToolbarItem>
                        ))}
                    </ToolbarGroup>
                    <ToolbarGroup>
                        <ToolbarItem className="pf-u-hidden-on-lg pf-u-mr-0">
                            <UserToggle isSmall extraItems={actions.map((action, key) => (
                                <DropdownItem key={key} component="button" isDisabled>{action.title}</DropdownItem>
                            ))} />
                        </ToolbarItem>
                        <ToolbarItem className="pf-u-screen-reader pf-u-visible-on-lg">
                            <UserToggle className='ins-c-dropdown__user'/>
                        </ToolbarItem>
                    </ToolbarGroup>
                    { isModalOpen && <InsightsAbout isModalOpen={isModalOpen} onClose={this.onModalToggle} /> }
                </Toolbar>
                <UserIcon/>
            </div>
        );
    }
}

export default Tools;
