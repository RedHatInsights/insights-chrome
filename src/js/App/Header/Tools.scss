.ins-c-toolbar__menu-user#UserMenu{
    display: flex;
}

.ins-c-toolbar__beta-badge {
    background-color: var(--pf-global--primary-color--200);
    margin-right: var(--pf-global--spacer--sm);
}
