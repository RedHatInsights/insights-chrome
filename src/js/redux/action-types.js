export const USER_LOGIN = '@@chrome/user-login';

export const APP_NAV_CLICK = '@@chrome/app-nav-click';

export const CHROME_PAGE_ACTION = '@@chrome/app-page-action';
export const CHROME_PAGE_OBJECT = '@@chrome/app-object-id';

export const CHROME_GET_ALL_TAGS = '@@chrome/get-all-tags';
export const CHROME_GET_ALL_SIDS = '@@chrome/get-all-sids';
export const CHROME_GET_ALL_WORKLOADS = '@@chrome/get-all-workloads';
export const GLOBAL_FILTER_SCOPE = '@@chrome/set-global-filter-scope';
export const GLOBAL_FILTER_UPDATE = '@@chrome/global-filter-update';
export const GLOBAL_FILTER_TOGGLE = '@@chrome/global-filter-toggle';
export const GLOBAL_FILTER_REMOVE = '@@chrome/global-filter-remove';

export const REGISTER_MODULE = '@@chrome/register-async-module';

export const TOGGLECONTEXTSWITCHER = '@@chrome/context-switcher-toggle';

export const LOAD_NAVIGATION_LANDING_PAGE = '@@chrome/load-navigation-landing-page';
export const LOAD_LEFT_NAVIGATION_SEGMENT = '@@chrome/load-navigation-segment';
export const LOAD_MODULES_SCHEMA = '@@chrome/load-modules-schema';

export const CHANGE_ACTIVE_MODULE = '@@chrome/change-active-module';
export const SET_PENDO_FEEDBACK_FLAG = '@@chrome/set-pendo-feedback-flag';
export const TOGGLE_FEEDBACK_MODAL = '@@chrome/toggle-feedback-modal';
export const UPDATE_ACCESS_REQUESTS_NOTIFICATIONS = '@@chrome/update-access-requests-notifications';
export const MARK_REQUEST_NOTIFICATION_SEEN = '@@chrome/mark-request-notification-seen';

export const STORE_INITIAL_HASH = '@@chrome/store-initial-hash';

export const POPULATE_QUICKSTARTS_CATALOG = '@@chrome/populate-quickstarts-catalog';
export const DISABLE_QUICKSTARTS = '@@chrome/disable-quickstarts';
