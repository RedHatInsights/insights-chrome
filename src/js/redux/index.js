import { applyReducerHash } from '@redhat-cloud-services/frontend-components-utilities/ReducerRegistry';

import {
  appNavClick,
  loginReducer,
  onPageAction,
  onPageObjectId,
  onRegisterModule,
  contextSwitcherBannerReducer,
  loadNavigationLandingPageReducer,
  loadNavigationSegmentReducer,
  loadModulesSchemaReducer,
  changeActiveModuleReducer,
  setPendoFeedbackFlag,
  toggleFeedbackModal,
  accessRequestsNotificationsReducer,
  markAccessRequestRequestReducer,
  storeInitialHashReducer,
  populateQuickstartsReducer,
  disableQuickstartsReducer,
} from './reducers';
import {
  onGetAllTags,
  onGetAllTagsPending,
  onSetGlobalFilterScope,
  onGlobalFilterToggle,
  onTagSelect,
  onGetAllSIDs,
  onGetAllSIDsPending,
  onGetAllWorkloads,
  onGetAllWorkloadsPending,
  onGlobalFilterRemove,
  globalFilterDefaultState,
} from './globalFilterReducers';
import {
  APP_NAV_CLICK,
  USER_LOGIN,
  CHROME_PAGE_ACTION,
  CHROME_PAGE_OBJECT,
  CHROME_GET_ALL_TAGS,
  CHROME_GET_ALL_SIDS,
  CHROME_GET_ALL_WORKLOADS,
  GLOBAL_FILTER_SCOPE,
  GLOBAL_FILTER_TOGGLE,
  GLOBAL_FILTER_UPDATE,
  GLOBAL_FILTER_REMOVE,
  REGISTER_MODULE,
  TOGGLECONTEXTSWITCHER,
  LOAD_NAVIGATION_LANDING_PAGE,
  LOAD_LEFT_NAVIGATION_SEGMENT,
  LOAD_MODULES_SCHEMA,
  CHANGE_ACTIVE_MODULE,
  SET_PENDO_FEEDBACK_FLAG,
  TOGGLE_FEEDBACK_MODAL,
  UPDATE_ACCESS_REQUESTS_NOTIFICATIONS,
  MARK_REQUEST_NOTIFICATION_SEEN,
  STORE_INITIAL_HASH,
  POPULATE_QUICKSTARTS_CATALOG,
  DISABLE_QUICKSTARTS,
} from './action-types';

const reducers = {
  [APP_NAV_CLICK]: appNavClick,
  [USER_LOGIN]: loginReducer,
  [CHROME_PAGE_ACTION]: onPageAction,
  [CHROME_PAGE_OBJECT]: onPageObjectId,
  [REGISTER_MODULE]: onRegisterModule,
  [TOGGLECONTEXTSWITCHER]: contextSwitcherBannerReducer,
  [LOAD_NAVIGATION_LANDING_PAGE]: loadNavigationLandingPageReducer,
  [LOAD_LEFT_NAVIGATION_SEGMENT]: loadNavigationSegmentReducer,
  [LOAD_MODULES_SCHEMA]: loadModulesSchemaReducer,
  [CHANGE_ACTIVE_MODULE]: changeActiveModuleReducer,
  [SET_PENDO_FEEDBACK_FLAG]: setPendoFeedbackFlag,
  [TOGGLE_FEEDBACK_MODAL]: toggleFeedbackModal,
  [UPDATE_ACCESS_REQUESTS_NOTIFICATIONS]: accessRequestsNotificationsReducer,
  [MARK_REQUEST_NOTIFICATION_SEEN]: markAccessRequestRequestReducer,
  [STORE_INITIAL_HASH]: storeInitialHashReducer,
  [POPULATE_QUICKSTARTS_CATALOG]: populateQuickstartsReducer,
  [DISABLE_QUICKSTARTS]: disableQuickstartsReducer,
};

const globalFilter = {
  [`${CHROME_GET_ALL_TAGS}_FULFILLED`]: onGetAllTags,
  [`${CHROME_GET_ALL_TAGS}_PENDING`]: onGetAllTagsPending,
  [`${CHROME_GET_ALL_SIDS}_FULFILLED`]: onGetAllSIDs,
  [`${CHROME_GET_ALL_SIDS}_PENDING`]: onGetAllSIDsPending,
  [`${CHROME_GET_ALL_WORKLOADS}_FULFILLED`]: onGetAllWorkloads,
  [`${CHROME_GET_ALL_WORKLOADS}_PENDING`]: onGetAllWorkloadsPending,
  [GLOBAL_FILTER_SCOPE]: onSetGlobalFilterScope,
  [GLOBAL_FILTER_TOGGLE]: onGlobalFilterToggle,
  [GLOBAL_FILTER_REMOVE]: onGlobalFilterRemove,
  [GLOBAL_FILTER_UPDATE]: onTagSelect,
};

export default function () {
  // const chromeInitialState = JSON.parse(localStorage.getItem('chrome')) || {};

  return {
    chrome: (
      state = {
        navigation: {},
        accessRequests: {
          count: 0,
          data: [],
        },
        quickstarts: {
          quickstarts: {},
        },
      },
      action
    ) => applyReducerHash(reducers)(state, action),
    globalFilter: (state = globalFilterDefaultState, action) => applyReducerHash(globalFilter)(state, action),
  };
}
