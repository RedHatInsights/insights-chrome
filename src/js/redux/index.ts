import { applyReducerHash } from '@redhat-cloud-services/frontend-components-utilities/ReducerRegistry';

import {
  accessRequestsNotificationsReducer,
  appNavClick,
  changeActiveModuleReducer,
  contextSwitcherBannerReducer,
  disableQuickstartsReducer,
  documentTitleReducer,
  loadModulesSchemaReducer,
  loadNavigationLandingPageReducer,
  loadNavigationSegmentReducer,
  loginReducer,
  markAccessRequestRequestReducer,
  markActiveProduct,
  onPageAction,
  onPageObjectId,
  onRegisterModule,
  populateQuickstartsReducer,
  setPendoFeedbackFlag,
  storeInitialHashReducer,
  toggleFeedbackModal,
} from './reducers';
import {
  globalFilterDefaultState,
  onGetAllSIDs,
  onGetAllSIDsPending,
  onGetAllTags,
  onGetAllTagsPending,
  onGetAllWorkloads,
  onGetAllWorkloadsPending,
  onGlobalFilterRemove,
  onGlobalFilterToggle,
  onSetGlobalFilterScope,
  onTagSelect,
} from './globalFilterReducers';
import {
  APP_NAV_CLICK,
  CHANGE_ACTIVE_MODULE,
  CHROME_GET_ALL_SIDS,
  CHROME_GET_ALL_TAGS,
  CHROME_GET_ALL_WORKLOADS,
  CHROME_PAGE_ACTION,
  CHROME_PAGE_OBJECT,
  DISABLE_QUICKSTARTS,
  GLOBAL_FILTER_REMOVE,
  GLOBAL_FILTER_SCOPE,
  GLOBAL_FILTER_TOGGLE,
  GLOBAL_FILTER_UPDATE,
  LOAD_LEFT_NAVIGATION_SEGMENT,
  LOAD_MODULES_SCHEMA,
  LOAD_NAVIGATION_LANDING_PAGE,
  MARK_ACTIVE_PRODUCT,
  MARK_REQUEST_NOTIFICATION_SEEN,
  POPULATE_QUICKSTARTS_CATALOG,
  REGISTER_MODULE,
  SET_PENDO_FEEDBACK_FLAG,
  STORE_INITIAL_HASH,
  TOGGLECONTEXTSWITCHER,
  TOGGLE_FEEDBACK_MODAL,
  UPDATE_ACCESS_REQUESTS_NOTIFICATIONS,
  UPDATE_DOCUMENT_TITLE_REDUCER,
  USER_LOGIN,
} from './action-types';
import { ChromeState, GlobalFilterState, ReduxState } from './store';
import { AnyAction } from 'redux';

const reducers = {
  [APP_NAV_CLICK]: appNavClick,
  [USER_LOGIN]: loginReducer,
  [CHROME_PAGE_ACTION]: onPageAction,
  [CHROME_PAGE_OBJECT]: onPageObjectId,
  [REGISTER_MODULE]: onRegisterModule,
  [TOGGLECONTEXTSWITCHER]: contextSwitcherBannerReducer,
  [LOAD_NAVIGATION_LANDING_PAGE]: loadNavigationLandingPageReducer,
  [LOAD_LEFT_NAVIGATION_SEGMENT]: loadNavigationSegmentReducer,
  [LOAD_MODULES_SCHEMA]: loadModulesSchemaReducer,
  [CHANGE_ACTIVE_MODULE]: changeActiveModuleReducer,
  [SET_PENDO_FEEDBACK_FLAG]: setPendoFeedbackFlag,
  [TOGGLE_FEEDBACK_MODAL]: toggleFeedbackModal,
  [UPDATE_ACCESS_REQUESTS_NOTIFICATIONS]: accessRequestsNotificationsReducer,
  [MARK_REQUEST_NOTIFICATION_SEEN]: markAccessRequestRequestReducer,
  [STORE_INITIAL_HASH]: storeInitialHashReducer,
  [POPULATE_QUICKSTARTS_CATALOG]: populateQuickstartsReducer,
  [DISABLE_QUICKSTARTS]: disableQuickstartsReducer,
  [UPDATE_DOCUMENT_TITLE_REDUCER]: documentTitleReducer,
  [MARK_ACTIVE_PRODUCT]: markActiveProduct,
};

const globalFilter = {
  [`${CHROME_GET_ALL_TAGS}_FULFILLED`]: onGetAllTags,
  [`${CHROME_GET_ALL_TAGS}_PENDING`]: onGetAllTagsPending,
  [`${CHROME_GET_ALL_SIDS}_FULFILLED`]: onGetAllSIDs,
  [`${CHROME_GET_ALL_SIDS}_PENDING`]: onGetAllSIDsPending,
  [`${CHROME_GET_ALL_WORKLOADS}_FULFILLED`]: onGetAllWorkloads,
  [`${CHROME_GET_ALL_WORKLOADS}_PENDING`]: onGetAllWorkloadsPending,
  [GLOBAL_FILTER_SCOPE]: onSetGlobalFilterScope,
  [GLOBAL_FILTER_TOGGLE]: onGlobalFilterToggle,
  [GLOBAL_FILTER_REMOVE]: onGlobalFilterRemove,
  [GLOBAL_FILTER_UPDATE]: onTagSelect,
};

export const chromeInitialState: ReduxState = {
  chrome: {
    contextSwitcherOpen: false,
    navigation: {},
    accessRequests: {
      hasUnseen: false,
      count: 0,
      data: [],
    },
    quickstarts: {
      quickstarts: {},
    },
    moduleRoutes: [],
  },
  globalFilter: {
    tags: {
      isLoaded: false,
      items: [],
    },
    workloads: {
      isLoaded: false,
    },
    sid: {
      isLoaded: false,
    },
    globalFilterHidden: false,
  },
};

export default function (): {
  chrome: (state: ChromeState, action: AnyAction) => ChromeState;
  globalFilter: (state: GlobalFilterState, action: AnyAction) => ChromeState;
} {
  // const chromeInitialState = JSON.parse(localStorage.getItem('chrome')) || {};

  return {
    chrome: (
      state = {
        navigation: {},
        accessRequests: {
          count: 0,
          data: [],
          hasUnseen: false,
        },
        quickstarts: {
          quickstarts: {},
        },
        contextSwitcherOpen: false,
        modules: {},
        scalprumConfig: {},
        moduleRoutes: [],
      },
      action
    ) => applyReducerHash(reducers)(state, action),
    globalFilter: (state = globalFilterDefaultState, action) => applyReducerHash(globalFilter)(state, action),
  };
}
