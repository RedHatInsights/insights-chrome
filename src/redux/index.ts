import { applyReducerHash } from '@redhat-cloud-services/frontend-components-utilities/ReducerRegistry';

import {
  globalFilterDefaultState,
  onGetAllSIDs,
  onGetAllSIDsPending,
  onGetAllTags,
  onGetAllTagsPending,
  onGetAllWorkloads,
  onGetAllWorkloadsPending,
  onGlobalFilterRemove,
  onGlobalFilterToggle,
  onSetGlobalFilterScope,
  onTagSelect,
} from './globalFilterReducers';
import {
  CHROME_GET_ALL_SIDS,
  CHROME_GET_ALL_TAGS,
  CHROME_GET_ALL_WORKLOADS,
  GLOBAL_FILTER_REMOVE,
  GLOBAL_FILTER_SCOPE,
  GLOBAL_FILTER_TOGGLE,
  GLOBAL_FILTER_UPDATE,
} from './action-types';
import { GlobalFilterState, ReduxState } from './store';
import { AnyAction } from 'redux';

const globalFilter = {
  [`${CHROME_GET_ALL_TAGS}_FULFILLED`]: onGetAllTags,
  [`${CHROME_GET_ALL_TAGS}_PENDING`]: onGetAllTagsPending,
  [`${CHROME_GET_ALL_SIDS}_FULFILLED`]: onGetAllSIDs,
  [`${CHROME_GET_ALL_SIDS}_PENDING`]: onGetAllSIDsPending,
  [`${CHROME_GET_ALL_WORKLOADS}_FULFILLED`]: onGetAllWorkloads,
  [`${CHROME_GET_ALL_WORKLOADS}_PENDING`]: onGetAllWorkloadsPending,
  [GLOBAL_FILTER_SCOPE]: onSetGlobalFilterScope,
  [GLOBAL_FILTER_TOGGLE]: onGlobalFilterToggle,
  [GLOBAL_FILTER_REMOVE]: onGlobalFilterRemove,
  [GLOBAL_FILTER_UPDATE]: onTagSelect,
};

export const chromeInitialState: ReduxState = {
  globalFilter: globalFilterDefaultState,
};

export default function (): {
  globalFilter: (state: GlobalFilterState, action: AnyAction) => GlobalFilterState;
} {
  return {
    globalFilter: (state = globalFilterDefaultState, action) => applyReducerHash(globalFilter)(state, action),
  };
}
