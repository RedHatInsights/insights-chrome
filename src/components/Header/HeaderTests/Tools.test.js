import React from 'react';
import Tools from '../Tools';
import { act, render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import { MemoryRouter } from 'react-router-dom';

jest.mock('../UserToggle', () => () => '<UserToggle />');
jest.mock('../ToolbarToggle', () => () => '<ToolbarToggle />');
jest.mock('../../../state/atoms/releaseAtom', () => {
  const util = jest.requireActual('../../../state/atoms/utils');
  return {
    __esModule: true,
    isPreviewAtom: util.atomWithToggle(false),
  };
});

jest.mock('@unleash/proxy-client-react', () => {
  const proxyClient = jest.requireActual('@unleash/proxy-client-react');
  return {
    __esModule: true,
    ...proxyClient,
    useFlag: () => {
      return true;
    },
  };
});

describe('Tools', () => {
  let assignMock = jest.fn();

  delete window.location;
  window.location = { assign: assignMock, href: '', pathname: '' };
  afterEach(() => {
    assignMock.mockClear();
  });
  it('should render correctly', async () => {
    const mockClick = jest.fn();
    let container;
    await act(async () => {
      container = render(
        <MemoryRouter>
          <Provider store={createStore((state = { chrome: { user: {}, notifications: { data: [] } } }) => state)}>
            <Tools onClick={mockClick} />
          </Provider>
        </MemoryRouter>
      ).container;
    });
    expect(container.querySelector('div')).toMatchSnapshot();
  });
});
