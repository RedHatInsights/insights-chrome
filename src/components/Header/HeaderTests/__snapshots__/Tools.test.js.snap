// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tools should render correctly 1`] = `
<div
  class="pf-c-toolbar__item pf-m-spacer-md"
>
  <label
    class="pf-c-switch pf-m-reverse chr-c-beta-switcher"
    data-ouia-component-id="OUIA-Generated-Switch-2"
    data-ouia-component-type="PF4/Switch"
    data-ouia-safe="true"
    for="reversed-switch"
  >
    <input
      aria-label="Preview switcher"
      class="pf-c-switch__input"
      id="reversed-switch"
      type="checkbox"
    />
    <span
      class="pf-c-switch__toggle"
    />
    <span
      aria-hidden="true"
      class="pf-c-switch__label pf-m-on"
    >
      Preview on
    </span>
    <span
      aria-hidden="true"
      class="pf-c-switch__label pf-m-off"
    >
      Preview off
    </span>
  </label>
</div>
`;
